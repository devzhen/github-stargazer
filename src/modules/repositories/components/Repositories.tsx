import * as R from 'ramda';
import React from 'react';
import { FlatList, TouchableOpacity, Image } from 'react-native';
import styled from 'styled-components/native';
import Swipeable from 'react-native-gesture-handler/Swipeable';

import Section from '@components/Section';

// @ts-ignore
import deleteIcon from './assets/ic_delete.png';

interface RepositoriesProps {
  repositories: any;
  isFetchingNow: boolean;
  onPressItem: any;
  deleteItem: any;
}

class Repositories extends React.Component<RepositoriesProps> {
  swipableRefs: any;

  constructor(props: any) {
    super(props);

    this.swipableRefs = {};
  }

  /**
   * Prepare repositories.
   */
  prepareRepositories = () => {
    const { repositories } = this.props;

    const prepared = (Object as any)
      .values(repositories)
      .sort((a: any, b: any) => {
        const aStar = R.propOr(0, 'stargazers_count', a);
        const bStar = R.propOr(0, 'stargazers_count', b);

        return bStar - aStar;
      });

    return prepared;
  };

  deleteItemHandler = (item: any) => {
    const { deleteItem } = this.props;

    const ref = R.prop(item.id, this.swipableRefs);
    const closeGesture = R.propOr(() => {}, 'close', ref);

    deleteItem(item);

    closeGesture();
  };

  /**
   * Key extractor.
   */
  keyExtractor = (item: any) => R.compose(R.toString, R.prop('id'))(item);

  /**
   * Render item.
   */
  renderItem = ({ item }: any) => {
    const { onPressItem } = this.props;

    const name = R.propOr('repo-name', 'name', item);
    const stargazersCount = R.compose((value: any) => {
      if (value > 1000) {
        return '1000+';
      }

      return value;
    }, R.propOr(0, 'stargazers_count'))(item);

    return (
      <Swipeable
        ref={(ref) => {
          this.swipableRefs = R.assoc(item.id, ref, this.swipableRefs);
        }}
        onSwipeableRightWillOpen={this.deleteItemHandler.bind(null, item)}
        renderRightActions={() => (
          <RemoveWrapper>
            <Image
              source={deleteIcon}
              style={{ height: 20, width: 20 }}
              resizeMode="contain"
            />
          </RemoveWrapper>
        )}
      >
        <TouchableOpacity onPress={onPressItem.bind(null, item)}>
          <Repository>
            <Card>
              <NameWrapper>
                <Name numberOfLines={1}>{name}</Name>
              </NameWrapper>
              <StarWrapper>
                <Stars>{stargazersCount}</Stars>
              </StarWrapper>
            </Card>
          </Repository>
        </TouchableOpacity>
      </Swipeable>
    );
  };

  render() {
    const { repositories, isFetchingNow } = this.props;

    if (isFetchingNow || R.isEmpty(repositories) || R.isNil(repositories)) {
      return null;
    }

    return (
      <Section fluid bgColor="#fff">
        <FlatListStyled
          data={this.prepareRepositories()}
          extraData={repositories}
          keyExtractor={this.keyExtractor}
          renderItem={this.renderItem}
          showsVerticalScrollIndicator={false}
          contentContainerStyle={{
            flexGrow: 1,
            paddingTop: 20,
            paddingBottom: 100,
          }}
        />
      </Section>
    );
  }
}

const FlatListStyled = styled(FlatList)`
  flex: 1;
`;

const Repository = styled.View`
  width: 100%;
  padding-horizontal: 20px;
  margin-bottom: 30px;
`;

const Card = styled.View`
  flex-direction: row;
  width: 100%;
  height: 50px;
  padding-left: 14px;
  border-width: 1px;
  border-color: ${(props) => props.theme.colors.gray};
  border-radius: 8px;
`;

const NameWrapper = styled.View`
  flex: 1;
  justify-content: center;
  align-items: flex-start;
`;

const StarWrapper = styled.View`
  justify-content: center;
  align-items: center;
  width: 50px;
  height: 100%;
`;

const Name = styled.Text`
  font-size: 17px;
  line-height: 20px;
  font-weight: 600;
  color: ${(props) => props.theme.colors.lightBlack};
`;

const Stars = styled(Name)`
  font-weight: normal;
  color: ${(props) => props.theme.colors.orange};
`;

const RemoveWrapper = styled.View`
  justify-content: center;
  align-items: center;
  width: 26px;
  height: 50px;
`;

export default Repositories;
