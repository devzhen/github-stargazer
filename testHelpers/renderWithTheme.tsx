import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';
import 'jest-styled-components';

import theme from '@theme/index';

const renderWithTheme = (component: any) =>
  renderer.create(<ThemeProvider theme={theme}>{component}</ThemeProvider>);

export default renderWithTheme;
