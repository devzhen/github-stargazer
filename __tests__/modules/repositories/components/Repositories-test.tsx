import React from 'react';
import 'jest-styled-components';

import Repositories from '@modules/repositories/components/Repositories';

// @ts-ignore
import renderWithTheme from '@testHelpers/renderWithTheme';

const repos = [
  { id: 1, name: '1' },
  { id: 2, name: '2' },
  { id: 3, name: '3' },
];

describe('Repositories repository component', () => {
  it('renders correctly, 1 case', () => {
    const tree = renderWithTheme(
      <Repositories
        isFetchingNow={true}
        repositories={[]}
        onPressItem={() => {}}
        deleteItem={() => {}}
      />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly, 2 case', () => {
    const tree = renderWithTheme(
      <Repositories
        isFetchingNow={false}
        repositories={repos}
        onPressItem={() => {}}
        deleteItem={() => {}}
      />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
