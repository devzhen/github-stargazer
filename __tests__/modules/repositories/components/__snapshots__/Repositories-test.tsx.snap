// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repositories repository component renders correctly, 1 case 1`] = `null`;

exports[`Repositories repository component renders correctly, 2 case 1`] = `
<View
  bgColor="#fff"
  borderColor={null}
  borderRadius={0}
  centerContentHorizontally={false}
  centerContentVertically={false}
  clientBottom={0}
  clientLeft={0}
  clientRight={0}
  clientTop={0}
  flexStartContentHorizontally={false}
  fluid={true}
  offsetBottom={0}
  offsetLeft={0}
  offsetRight={0}
  offsetTop={0}
  stretchContent={false}
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flexBasis": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "marginBottom": 0,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
        "paddingBottom": 0,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 0,
      },
      Object {},
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "paddingBottom": 100,
        "paddingTop": 20,
      }
    }
    data={
      Array [
        Object {
          "id": 1,
          "name": "1",
        },
        Object {
          "id": 2,
          "name": "2",
        },
        Object {
          "id": 3,
          "name": "3",
        },
      ]
    }
    disableVirtualization={false}
    extraData={
      Array [
        Object {
          "id": 1,
          "name": "1",
        },
        Object {
          "id": 2,
          "name": "2",
        },
        Object {
          "id": 3,
          "name": "3",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          collapsable={false}
          forwardedRef={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onLayout={[Function]}
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "flexDirection": "row-reverse",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateX": -10000,
                  },
                ],
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "height": 50,
                    "justifyContent": "center",
                    "width": 26,
                  },
                ]
              }
            >
              <Image
                resizeMode="contain"
                source={
                  Object {
                    "testUri": "../../../src/modules/repositories/components/assets/ic_delete.png",
                  }
                }
                style={
                  Object {
                    "height": 20,
                    "width": 20,
                  }
                }
              />
            </View>
            <View
              onLayout={[Function]}
            />
          </View>
          <View
            collapsable={false}
            forwardedRef={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            pointerEvents="auto"
            style={
              Object {
                "transform": Array [
                  Object {
                    "translateX": -0,
                  },
                ],
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "marginBottom": 30,
                      "paddingHorizontal": 20,
                      "width": "100%",
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "borderColor": "#d3d3d3",
                        "borderRadius": 8,
                        "borderWidth": 1,
                        "flexDirection": "row",
                        "height": 50,
                        "paddingLeft": 14,
                        "width": "100%",
                      },
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "flex-start",
                          "flexBasis": 0,
                          "flexGrow": 1,
                          "flexShrink": 1,
                          "justifyContent": "center",
                        },
                      ]
                    }
                  >
                    <Text
                      numberOfLines={1}
                      style={
                        Array [
                          Object {
                            "color": "#383838",
                            "fontSize": 17,
                            "fontWeight": "600",
                            "lineHeight": 20,
                          },
                        ]
                      }
                    >
                      1
                    </Text>
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "height": "100%",
                          "justifyContent": "center",
                          "width": 50,
                        },
                      ]
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#eb9e38",
                            "fontSize": 17,
                            "fontWeight": "normal",
                            "lineHeight": 20,
                          },
                        ]
                      }
                    >
                      0
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          collapsable={false}
          forwardedRef={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onLayout={[Function]}
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "flexDirection": "row-reverse",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateX": -10000,
                  },
                ],
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "height": 50,
                    "justifyContent": "center",
                    "width": 26,
                  },
                ]
              }
            >
              <Image
                resizeMode="contain"
                source={
                  Object {
                    "testUri": "../../../src/modules/repositories/components/assets/ic_delete.png",
                  }
                }
                style={
                  Object {
                    "height": 20,
                    "width": 20,
                  }
                }
              />
            </View>
            <View
              onLayout={[Function]}
            />
          </View>
          <View
            collapsable={false}
            forwardedRef={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            pointerEvents="auto"
            style={
              Object {
                "transform": Array [
                  Object {
                    "translateX": -0,
                  },
                ],
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "marginBottom": 30,
                      "paddingHorizontal": 20,
                      "width": "100%",
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "borderColor": "#d3d3d3",
                        "borderRadius": 8,
                        "borderWidth": 1,
                        "flexDirection": "row",
                        "height": 50,
                        "paddingLeft": 14,
                        "width": "100%",
                      },
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "flex-start",
                          "flexBasis": 0,
                          "flexGrow": 1,
                          "flexShrink": 1,
                          "justifyContent": "center",
                        },
                      ]
                    }
                  >
                    <Text
                      numberOfLines={1}
                      style={
                        Array [
                          Object {
                            "color": "#383838",
                            "fontSize": 17,
                            "fontWeight": "600",
                            "lineHeight": 20,
                          },
                        ]
                      }
                    >
                      2
                    </Text>
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "height": "100%",
                          "justifyContent": "center",
                          "width": 50,
                        },
                      ]
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#eb9e38",
                            "fontSize": 17,
                            "fontWeight": "normal",
                            "lineHeight": 20,
                          },
                        ]
                      }
                    >
                      0
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          collapsable={false}
          forwardedRef={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onLayout={[Function]}
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "flexDirection": "row-reverse",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateX": -10000,
                  },
                ],
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "height": 50,
                    "justifyContent": "center",
                    "width": 26,
                  },
                ]
              }
            >
              <Image
                resizeMode="contain"
                source={
                  Object {
                    "testUri": "../../../src/modules/repositories/components/assets/ic_delete.png",
                  }
                }
                style={
                  Object {
                    "height": 20,
                    "width": 20,
                  }
                }
              />
            </View>
            <View
              onLayout={[Function]}
            />
          </View>
          <View
            collapsable={false}
            forwardedRef={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            pointerEvents="auto"
            style={
              Object {
                "transform": Array [
                  Object {
                    "translateX": -0,
                  },
                ],
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "marginBottom": 30,
                      "paddingHorizontal": 20,
                      "width": "100%",
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "borderColor": "#d3d3d3",
                        "borderRadius": 8,
                        "borderWidth": 1,
                        "flexDirection": "row",
                        "height": 50,
                        "paddingLeft": 14,
                        "width": "100%",
                      },
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "flex-start",
                          "flexBasis": 0,
                          "flexGrow": 1,
                          "flexShrink": 1,
                          "justifyContent": "center",
                        },
                      ]
                    }
                  >
                    <Text
                      numberOfLines={1}
                      style={
                        Array [
                          Object {
                            "color": "#383838",
                            "fontSize": 17,
                            "fontWeight": "600",
                            "lineHeight": 20,
                          },
                        ]
                      }
                    >
                      3
                    </Text>
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "height": "100%",
                          "justifyContent": "center",
                          "width": 50,
                        },
                      ]
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#eb9e38",
                            "fontSize": 17,
                            "fontWeight": "normal",
                            "lineHeight": 20,
                          },
                        ]
                      }
                    >
                      0
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
